[Jetpack] Add Jetpack powered banner to Reader and limit a11y scaling

twstokes opened the pull request:
## Description

This PR adds the "Jetpack powered" banner to the following areas in Reader:
- Following tab
- Likes tab
- Saved tab
- Search -> Posts

It also tweaks accessibility sizes for banners and badges by limiting the font scaling to 22pts. Also, the Jetpack logo no longer scales dynamically.

**Note:** If we feel the banner can get _too_ large we should be able to easily limit this by setting a max height constraint.

## Testing

### Reader banner
1. Open the WordPress app
2. Tap Reader
3. Observe the "Jetpack powered" banner in these areas:
- Following tab
- Likes tab
- Saved tab
- Search -> Posts

| Reader banner iPhone | Reader banner iPad |
| - | - |
| ![Simulator Screen Shot - iPhone 13 - 2022-07-18 at 09 37 46](https://user-images.githubusercontent.com/2092798/179533151-cf2f5355-4340-4608-ada7-2222ce4e103e.png) | ![Simulator Screen Shot - iPad (9th generation) - 2022-07-18 at 10 26 18](https://user-images.githubusercontent.com/2092798/179533465-9d188a64-5e1d-45ce-8244-953184bf3d00.png) |

### Reader banner animations
1. Open the WordPress app
2. Tap Reader
3. Scroll in the Reader view
4. Expect the banner to follow the behavior [defined in this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/19040)

<img src="https://user-images.githubusercontent.com/2092798/179568737-20aac273-1ada-4a08-9224-3ae856dd1a20.gif" width="350px" />

### Accessibility sizes

| Normal banner | Largest a11y size |
| - | - |
| ![Simulator Screen Shot - iPhone 13 - 2022-07-18 at 12 03 25](https://user-images.githubusercontent.com/2092798/179555641-799d5bd1-289d-49eb-8ad9-713315bf8d7b.png) | ![Simulator Screen Shot - iPhone 13 - 2022-07-18 at 12 02 45](https://user-images.githubusercontent.com/2092798/179555744-c641e9a2-fd2f-49a0-ab4f-611417e87ecf.png) |

| Normal badge | Large a11y size |
| - | - |
| ![Simulator Screen Shot - iPhone 13 - 2022-07-18 at 12 15 29](https://user-images.githubusercontent.com/2092798/179556120-8ab9227b-bcfb-40b2-9421-27e35037a7fa.png) | ![Simulator Screen Shot - iPhone 13 - 2022-07-18 at 12 15 41](https://user-images.githubusercontent.com/2092798/179556147-02073fe9-9073-484a-b153-7b832aeeac1d.png) |

#### Stats
1. Open the WordPress app
6. On the My Site view, go to Stats
7. Observe the "Jetpack powered" banner
8. Increase the device's font size to large accessibility settings
9. Expect the banner to render correctly

#### Notifications
1. Open the WordPress app
2. Tap Notifications
3. Observe the "Jetpack powered" banner
4. Increase the device's font size to large accessibility settings
5. Expect the banner to render correctly

### Items to test:
- Large accessibility sizes
- Dark mode
- Landscape orientation
- iPad
- Banner should not appear in Jetpack app
- Banner should not appear when the feature flag is disabled

## Regression Notes
1. Potential unintended areas of impact
    - Reader functionality
    - Stats banner
    - Notifications banner
    - Badges (My Site -> Activity Log)

2. What I did to test those areas of impact (or what existing automated tests I relied on)
    - Manual Reader tests

3. What automated tests I added (or what prevented me from doing so)
    - None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

twstokes committed:
Add Jetpack powered banner to ReaderStreamViewController.

twstokes committed:
Allow Jetpack banner image to scale with a11y sizes.

twstokes committed:
Let Stats and Notifications Jetpack banner grow with larger accessibility sizes.

twstokes commented:
Going back to draft mode while I tweak the max scaled size of the banner.

shaunandrews commented:
I did consider the scaling when designing the banner and badge. I'm not very familiar with the best practices here, but a quick Google tells me that not all elements need to increase their size. I think it would be good to cap the size of the banner and badge.

<img width="1556" alt="image" src="https://user-images.githubusercontent.com/191598/179548734-30f802b0-3021-4776-8d81-9543c3174d07.png">

In the above image, I increased the font-size (18px -> 22px) for both the banner and the badge. I also added an extra 5px to the top/bottom padding on the banner. I'm not entirely sure how to best represent this, but hopefully this helps.

twstokes committed:
Don't scale the Jetpack logo and limit the a11y font scaling for the banner and badge.

twstokes commented:
> I did consider the scaling when designing the banner and badge. I'm not very familiar with the best practices here, but a quick Google tells me that not all elements need to increase their size. I think it would be good to cap the size of the banner and badge.
> 
> <img alt="image" width="1556" src="https://user-images.githubusercontent.com/191598/179548734-30f802b0-3021-4776-8d81-9543c3174d07.png">
> 
> In the above image, I increased the font-size (18px -> 22px) for both the banner and the badge. I also added an extra 5px to the top/bottom padding on the banner. I'm not entirely sure how to best represent this, but hopefully this helps.

Thanks @shaunandrews. I've added screenshots comparing the normal and scaled sizes, which are capped at 22pts.

twstokes committed:
Merge remote-tracking branch 'origin/trunk' into task/reader-jetpack-powered

twstokes committed:
Fix animation layout issue where view would teleport to the top of the stack view.

twstokes committed:
Conform to JPScrollViewDelegate and send scroll view events.

twstokes commented:
This is now ready for review. 🙇

Gio2018 commented on WordPress/Classes/ViewRelated/Jetpack/Branding/Banner/JetpackBannerView.swift:
```diff
@@ -47,6 +47,7 @@ extension JetpackBannerView: Subscriber {
 
         UIView.animate(withDuration: 0.3, delay: 0, options: [.curveEaseIn], animations: { [weak self] in
             self?.isHidden = isHidden
+            self?.superview?.layoutIfNeeded()
```
🎉 🎉

Gio2018 commented on WordPress/Classes/ViewRelated/Jetpack/Branding/Button/JetpackButton.swift:
```diff
@@ -74,7 +74,7 @@ class JetpackButton: UIButton {
         backgroundColor = buttonBackgroundColor
         setTitleColor(buttonTitleColor, for: .normal)
-        titleLabel?.font = Appearance.titleFont
+        titleLabel?.font = UIFontMetrics.default.scaledFont(for: Appearance.titleFont, maximumPointSize: 22)
```
nit: can we put this whole thing inside Appearance.titleFont?

twstokes replied:
Good call - I'd missed that it was local to the button and not the app. 👍

twstokes replied:
This should be resolved with 72434a5da17fd28bd979ac2141ac14b8af2a1e3b. Thanks!

Gio2018 commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -483,28 +486,44 @@ import WordPressFlux
+        }
+
+        let jetpackBannerView = JetpackBannerView()
+        jetpackBannerView.setup()
```
I think I should've set the `private` access modifier on that `setup()` method in `JetpackBannerView`. 
This way we could simply call here `JetpackBannerView(frame: .zero)` and then maybe add  `jetpackBannerView.translateAutoresizingMaskIntoConstraints = false` . What do you think?

twstokes replied:
Ah yes, I'll tweak this. 👍

twstokes replied:
This should be resolved with 21479881509677ca1f1b5fc475cf523352e0b2b1. Thanks!

Gio2018 commented on WordPress/Classes/ViewRelated/Reader/ReaderStreamViewController.swift:
```diff
@@ -483,28 +486,44 @@ import WordPressFlux
+        jetpackBannerView.setup()
+        addTranslationObserver(jetpackBannerView)
+        stackView.addArrangedSubview(jetpackBannerView)
+        jetpackBannerView.heightAnchor.constraint(greaterThanOrEqualToConstant: 50).isActive = true
```
Normally I would be bitching about putting the `50` constant in a group of constants 🤣, but looking at this file, it would probably just complicate it further.

twstokes replied:
@Gio2018 what about making it a static var for the JetpackBannerView? e749c114ea7099ee4fded14a81d624cab06a9810

Gio2018 replied:
That sounds good to me!

Gio2018 reviewed (approved):
Worked well for me @twstokes !
I left a couple of optional comments, but other than that :shipit: !

twstokes committed:
Make JetpackBannerView setup private and don't double-call.

twstokes committed:
Move font settings to static var.

twstokes committed:
Add static var for JetpackBannerView minimum height.
